### DISCUSSION TITLE
Is it possible to use the EventBus outside of a server?

### DISCUSSION THREAD

**Discussion by matthewpflueger** [marblejs/marble#379] [Q&A]
I would like to use the local only EventBus and client to decouple some client-side components in the browser. Is it possible to use the EventBus in this manner? 

I ask because all the examples show using the EventBus with a server but looking at the code it doesn't seem like it would be a stretch for it to operate outside of a server (the test bed for it seems to do just that).

**Comment by JozefFlakus** [comment]
Hi @matthewpflueger,

Good question. I think it could be possible to achieve with version 4.0 which decouples HTTP module from the `@marblejs/core` to the separate package. To be honest, didn't test it in this context. 🤔 Theoretically `EventBus` is just a variation on the `redux-observable` library.

**Reply by drwlrsn** [reply]
Oh I have wished the same thing. I quite like the EventBus and want to use it in a vanilla Node + Express app. I didn't have the opportunity to choose marble.js at the start, but this could be the next best thing.

**Reply by JozefFlakus** [reply]
@drwlrsn In this case I think it is not a problem. You just need a `@marblejs/core` + `@marblejs/messaging`.

Here you have some handy snippets:

```typescript
import { firstValueFrom } from 'rxjs';
import { pipe } from 'fp-ts/lib/function';
import { contextFactory, bindEagerlyTo, useContext, lookup, Event } from '@marblejs/core';
import { messagingListener, MessagingListenerConfig, EventBusToken, EventBusClientToken } from '@marblejs/messaging';

export const bootstrapEventBus = async (config: MessagingListenerConfig) =>
  pipe(
    await contextFactory(
      bindEagerlyTo(EventBusToken)(EventBus({ listener: messagingListener(config) })),
      bindEagerlyTo(EventBusClientToken)(EventBusClient),
    ),
    lookup,
    ask => ({
      eventBus: useContext(EventBusToken)(ask),
      eventBusClient: useContext(EventBusClientToken)(ask),
    } as const),
  );

// then you can send messages to the main EventBus
const someEvent: Event = { type: 'FOO', payload: 'bar' };
// according to the docs, please remember that `send` returns an Observable
// which means you have to subscribe to it or trigger it like this
await firstValueFrom(eventBusClient.send(event));

// when done always remember to close a opened EventBus connection (internally: RxJS subscribers)
await eventBus.close();
await eventBusClient.close();
```

I hope it will help you somehow :)

For more details please follow this file:
https://github.com/marblejs/marble/blob/master/packages/messaging/src/eventbus/messaging.eventBus.reader.spec.ts

**Reply by drwlrsn** [reply]
That's dope. Thanks. I use marble.js on a side project and it's been a great way to apply FP and learn about reactive programming. Thanks for all the work you've done.
